app/oidc/gcp_oidc.py

```diff
@@ -0,0 +1,35 @@
+from cachetools.func import ttl_cache
+from google.auth.credentials import Credentials
+from google.auth.transport.requests import Request
+from google.oauth2.id_token import fetch_id_token_credentials
+from structlog import get_logger
+
+from app.oidc.oidc import OIDCCredentialsService
+from app.settings import OIDC_TOKEN_LEEWAY_IN_SECONDS, OIDC_TOKEN_VALIDITY_IN_SECONDS
+
+TTL = OIDC_TOKEN_VALIDITY_IN_SECONDS - OIDC_TOKEN_LEEWAY_IN_SECONDS
+
+logger = get_logger()
+
+
+class OIDCCredentialsServiceGCP(OIDCCredentialsService):
+    @staticmethod
+    @ttl_cache(maxsize=None, ttl=TTL)
+    def get_credentials(*, iap_client_id: str) -> Credentials:
```

Comment:
Something to consider is a retry, we pretty much retry all our external requests.

Reply:
You can pass in a session object to the `Request(session=session)`. Here you could reuse our `get_retryable_session()` method and configure it appropriately. I haven't tested if this works out of the box.

For now, I'm happy for us to reuse:

```py
SUPPLEMENTARY_DATA_REQUEST_BACKOFF_FACTOR = 0.2
SUPPLEMENTARY_DATA_REQUEST_MAX_RETRIES = 2  # Totals no. of request should be 3. The initial request + SUPPLEMENTARY_DATA_REQUEST_MAX_RETRIES
SUPPLEMENTARY_DATA_REQUEST_TIMEOUT = 3
SUPPLEMENTARY_DATA_REQUEST_RETRY_STATUS_CODES = [
    408,
    429,
    500,
    502,
    503,
    504,
]
```

We can tune later.

Reply:
Can we check if Timeout (should be) and Connection errors are retried.

Reply:
have done some digging and the google auth library has a retry built into it, in each case it ends up in `_token_endpoint_request_no_throw` and - if the initial request fails it goes into this
```
retries = _exponential_backoff.ExponentialBackoff()
    for _ in retries:
        request_succeeded, response_data, retryable_error = _perform_request()
        if request_succeeded or not retryable_error:
            return request_succeeded, response_data, retryable_error

    return False, response_data, retryable_error
```
which by default attempts 3 times with an exponential backoff iterator

So I am thinking we probably don't need to add our own on top?

Reply:
`retryable_error` will return true if it errors and the status code is internal error, service unavailable or request timeout, so we should be covered on the necessary ones

Reply:
For context for other, the above is not used when calling the metadata server, but it does look like GCP does retry up to 5 times when calling the metadata server. It does seem to only retry Transport errors which seems appropriate given it is not an external call. Therefore, as discussed, just some tests to verify the retry is working.

Reply:
I've pushed up an attempt at this - I'm not sure how much value we're going to get from this test given the lib should have tests and the amount I ended up needing to mock to get this to work
or maybe I've missed something obvious 😄 
is this on the lines you were thinking or did you have a different idea of how it would look?

Reply:
I suppose its only 4 patches which isn't too bad

Reply:
Will have a look, but the purpose was mainly to ensure that there is a retry than retry works. The lib tests would check it, but our test would fail if Google were to remove the retry, i.e lets say they expect users to pass in a retry strategy. We had similar issues with GCS.

Reply:
Ah cool - makes sense, this should cover it having the retries

Reply:
LGTM.